src/pczt.rs

```diff
@@ -0,0 +1,294 @@
+
+    /// The value commitment randomness.
+    ///
+    /// - This is set by the Constructor.
+    /// - The IO Finalizer compresses it into the `bsk`.
+    /// - This is required by the Prover.
+    /// - This may be used by Signers to verify that the value correctly matches `cv`.
+    ///
+    /// This opens `cv` for all participants. For Signers who don't need this information,
+    /// or after proofs / signatures have been applied, this can be redacted.
+    pub(crate) rcv: Option<ValueCommitTrapdoor>,
+
+    /// The proof generation key `(ak, nsk)` corresponding to the recipient that received
+    /// the note being spent.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    pub(crate) proof_generation_key: Option<ProofGenerationKey>,
+
+    /// A witness from the note to the bundle's anchor.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    pub(crate) witness: Option<MerklePath>,
+
+    /// The spend authorization randomizer.
+    ///
+    /// - This is chosen by the Constructor.
+    /// - This is required by the Signer for creating `spend_auth_sig`, and may be used to
+    ///   validate `rk`.
+    /// - After`zkproof` / `spend_auth_sig` has been set, this can be redacted.
+    pub(crate) alpha: Option<jubjub::Scalar>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the note
+    /// being spent.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// The spend authorizing key for this spent note, if it is a dummy note.
+    ///
+    /// - This is chosen by the Constructor.
+    /// - This is required by the IO Finalizer, and is cleared by it once used.
+    /// - Signers MUST reject PCZTs that contain `dummy_ask` values.
+    pub(crate) dummy_ask: Option<SpendAuthorizingKey>,
+
+    /// Proprietary fields related to the note being spent.
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+/// Information about a Sapling output within a transaction.
+///
+/// This struct is for representing Sapling outputs in a partially-created transaction. If
+/// you have a fully-created transaction, use [the regular `OutputDescription` struct].
+///
+/// [the regular `OutputDescription` struct]: crate::bundle::OutputDescription
+#[derive(Getters)]
+#[getset(get = "pub")]
+pub struct Output {
+    /// A commitment to the value created by this output.
+    pub(crate) cv: ValueCommitment,
+
+    /// A commitment to the new note being created.
+    pub(crate) cmu: ExtractedNoteCommitment,
+
+    /// The ephemeral key used to encrypt the note plaintext.
+    pub(crate) ephemeral_key: EphemeralKeyBytes,
+
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Once we have memo bundles, we will be able to set memos independently of Outputs.
+    /// For now, the Constructor sets both at the same time.
+    pub(crate) enc_ciphertext: [u8; ENC_CIPHERTEXT_SIZE],
```

Comment:
Should `Output` be polymorphic in the type here just so that the type doesn't have to change with ZIP 231?

Reply:
I don't have the time to design that such that I can ensure it's forward compatible. And we'll need to change much of the rest of the API anyway, so this isn't much extra.

Reply:
For Orchard it's a `Vec<u8>`. Isn't it better to do the same here for consistency?

Reply:
[It already does](https://github.com/zcash/librustzcash/blob/205d4c930319b7b6d24aeb4efde69e9b4d1b6f7b/pczt/src/sapling.rs#L165-L177).